### PR TITLE
[WIP] Auto merge region 

### PR DESCRIPTION
**Pull request by visualYJD**

### CODE REVIEW

File: src/merge/merge_checker.cc [hunk 1]
```diff
@@ -0,0 +1,560 @@
...
+      merge_to_region->InnerRegion().definition().index_parameter().has_vector_index_parameter()) {
+    merge_to_is_vector_index = true;
+  }
+
+  if (merge_from_is_vector_index != merge_to_is_vector_index) {
+    need_scan_check = false;
+    reason = "merge_from_region and merge_to_region has different region type";
+    return need_scan_check;
+  }
+  if (merge_from_is_vector_index && merge_to_is_vector_index && !merge_from_region->VectorIndexWrapper()->IsReady()) {
```
Comment (rock-git) [inline src/merge/merge_checker.cc:111]: 这里也没必要吧

File: src/merge/merge_checker.cc [hunk 2]
```diff
@@ -0,0 +1,560 @@
...
+    if (size > 0) {
+      region_metrics->SetRegionSize(size);
+    }
+  }
+
+  DINGO_LOG(INFO) << fmt::format("[merge.check][region({})] update actual size({}), count({}) elapsed time({}ms)",
+                                 region->Id(), region_size, count, Helper::TimestampMs() - start_time);
+  return butil::Status::OK();
+}
+void PreMergeCheckTask::PreMergeCheck() {
```
Comment (rock-git) [inline src/merge/merge_checker.cc:397]: 函数直接通过空行隔开

File: src/merge/merge_checker.cc [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+
+  int64_t size = 0;
+  int64_t count = 0;
+  for (; iter.Valid(); iter.Next()) {
+    size += iter.KeyValueSize();
+    count++;
+  }
+  region_size = size;
+  reigon_keys_count = count;
+  // Update region key count metrics.
```
Comment (rock-git) [inline src/merge/merge_checker.cc:382]: 还有size

File: src/merge/merge_checker.cc [hunk 4]
```diff
@@ -0,0 +1,560 @@
...
+      }
+      if (merge_to->LastSplitTimestamp() != 0 &&
+          merge_to->LastSplitTimestamp() + interval * 1000 < Helper::TimestampMs()) {
+        need_scan_check = false;
+        reason = "merge to region recently split ";
+        break;
+      }
+      if (merge_from->Definition().range().end_key() != merge_to->Definition().range().start_key()) {
+        need_scan_check = false;
+        reason = "merge_from_region and merge_to_region has not continuous range";
```
Comment (rock-git) [inline src/merge/merge_checker.cc:481]: 简化region range has not continuous

File: src/merge/merge_checker.cc [hunk 5]
```diff
@@ -0,0 +1,560 @@
...
+        break;
+      }
+      // validate region store engine
+      if (merge_from->Definition().store_engine() != merge_to->Definition().store_engine()) {
+        need_scan_check = false;
+        reason = "merge_from_region and merge_to_region has different store_engine";
```
Comment (rock-git) [inline src/merge/merge_checker.cc:487]: 简化store_engine is different

File: src/merge/merge_checker.cc [hunk 6]
```diff
@@ -0,0 +1,560 @@
...
+        break;
+      }
+      // validate region raw engine
+      if (merge_from->Definition().raw_engine() != merge_to->Definition().raw_engine()) {
+        need_scan_check = false;
+        reason = "merge_from_region and merge_to_region has different raw_engine";
```
Comment (rock-git) [inline src/merge/merge_checker.cc:493]: 简化 raw_engine is different

File: src/merge/merge_checker.cc [hunk 7]
```diff
@@ -0,0 +1,560 @@
...
+        break;
+      }
+      // validate region part id
+      if (merge_from->Definition().part_id() != merge_to->Definition().part_id()) {
+        need_scan_check = false;
+        reason = "merge_from_region and merge_to_region has different part_id";
```
Comment (rock-git) [inline src/merge/merge_checker.cc:499]: 简化 region partition is different

File: src/merge/merge_checker.cc [hunk 8]
```diff
@@ -0,0 +1,560 @@
...
+        break;
+      }
+      // validate region peers
+      if (Helper::IsDifferencePeers(merge_from->Definition(), merge_to->Definition())) {
+        need_scan_check = false;
+        reason = "merge_from_region and merge_to_region peers is differencce.";
```
Comment (rock-git) [inline src/merge/merge_checker.cc:505]: 简化 region peers is different

File: src/merge/merge_checker.cc [hunk 9]
```diff
@@ -0,0 +1,560 @@
...
+      }
+      if (CheckLeaderAndFollowerStatus(merge_from->Id())) {
+        need_scan_check = false;
+        reason = "not leader or follower abnormal";
+        break;
+      }
+      if (merge_from->LastSplitTimestamp() != 0 &&
+          merge_from->LastSplitTimestamp() + interval * 1000 < Helper::TimestampMs()) {
+        need_scan_check = false;
+        reason = "merge from region recently split ";
```
Comment (rock-git) [inline src/merge/merge_checker.cc:470]: 简化region recently has split

File: src/merge/merge_checker.cc [hunk 10]
```diff
@@ -0,0 +1,560 @@
...
+        break;
+      }
+      if (merge_to->LastSplitTimestamp() != 0 &&
+          merge_to->LastSplitTimestamp() + interval * 1000 < Helper::TimestampMs()) {
+        need_scan_check = false;
+        reason = "merge to region recently split ";
```
Comment (rock-git) [inline src/merge/merge_checker.cc:476]: 简化region recently  has split

File: src/merge/merge_checker.h [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+      : merge_check_workers_(merge_check_workers),
+        merge_from_region_(merge_from_region),
+        merge_to_region_(merge_to_region) {}
+  ~MergeCheckTask() override = default;
+
+  std::string Type() override { return "MERGE_CHECK"; }
+
+  void Run() override {
+    MergeCheck();
+    if (merge_from_region_ != nullptr && merge_to_region_ != nullptr && merge_check_workers_ != nullptr) {
```
Comment (rock-git) [inline src/merge/merge_checker.h:73]: 可以不用检查，这个类的入口校验

File: src/merge/merge_checker.cc [hunk 12]
```diff
@@ -0,0 +1,560 @@
...
+                                   need_scan_check, reason);
+    if (!need_scan_check) {
+      continue;
+    }
+    checking_regions.insert(merge_from->Id());
+    checking_regions.insert(merge_to->Id());
+
+    auto task = std::make_shared<MergeCheckTask>(merge_check_workers_, merge_from, merge_to);
+    if (merge_check_workers_->Execute(task)) {
+      merge_check_workers_->AddRegionChecking(merge_from->Id());
```
Comment (rock-git) [inline src/merge/merge_checker.cc:529]: 放这里不合适，有可能任务已经执行完了

File: conf/store-gflags.conf [hunk 13]
```diff
@@ -7,7 +7,7 @@
 -free_memory_to_system_interval=60
 -enable_dir_service=false
 -enable_threads_service=false
--dingo_log_switch_txn_detail=false
+-dingo_log_switch_txn_detail=true
```
[the remark this answers is not in the document]
Comment (visualYJD) [reply conf/store-gflags.conf:10]: 这里不需要开启，我还原一下这里。

**Review by rock-git — APPROVED** [review]
LGTM

**Pull request merged**